File: 695-max-area-of-island/695-max-area-of-island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        def bfs(i_orig, j_orig) -> int:
            if not grid[i_orig][j_orig]: return 0
            q = deque([(i_orig, j_orig)])
            size = 0
            while q:
                i, j = q.popleft()
                if not visited[i][j]:
                    if i - 1 >= 0 and grid[i - 1][j]: q.append((i - 1, j))
                    if i + 1 < len(grid) and grid[i + 1][j]: q.append((i + 1, j))
                    if j - 1 >= 0 and grid[i][j - 1]: q.append((i, j - 1))
                    if j + 1 < len(grid[0]) and grid[i][j + 1]: q.append((i, j + 1))
                    size += 1
                visited[i][j] = 1
            return size
        
        visited = [[0] * len(grid[0]) for _ in range(len(grid))]
        max_size = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if visited[i][j]: continue
                size = bfs(i, j)
                max_size = max(max_size, size)
        return max_size

File: 695-max-area-of-island/695-max-area-of-island.cpp
class Solution {
    int dfs(vector<vector<int>>& grid, int i, int j) {
        if (!grid[i][j]) return 0;
        grid[i][j] = 0;
        int total = 1;
        if (i > 0) total += dfs(grid, i - 1, j);
        if (j > 0) total += dfs(grid, i, j - 1);
        if (i < grid.size() - 1) total += dfs(grid, i + 1, j);
        if (j < grid[0].size() - 1) total += dfs(grid, i, j + 1);
        return total;
    }
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        ios_base::sync_with_stdio(0);
        cin.tie(0);
        cout.tie(0);
        
        int maxArea = 0;
        for (int i = 0; i < grid.size(); ++i)
            for (int j = 0; j < grid[0].size(); ++j)
                if (grid[i][j]) maxArea = max(maxArea, dfs(grid, i, j));
        return maxArea;
    }
};